sdks/go/pkg/beam/testing/passert/equals.go

```diff
@@ -53,13 +54,37 @@ func EqualsList(s beam.Scope, col beam.PCollection, list interface{}) beam.PColl
 	return equals(subScope, col, listCollection)
 }
 
+// WindowedEqualsList verifies that the given collection has the same values as a
+// given list, under coder equality. The values must be provided as an array or a slice.
+// This function also takes a window function to window the list into for cases where
+// the PCollections cannot be globally windowed (e.g. tests in unbounded pipelines.)
+// This windowing function is applied to both the PCollection created from the list
+// and the impulse used to trigger the Diff function.
+func WindowedEqualsList(s beam.Scope, wfn *window.Fn, col beam.PCollection, list interface{}) beam.PCollection {
```

Comment:
I think I'm on board with adding this function (it also just seems generally useful to be able to test any custom windows users might eventually have). I'm wondering, however, if it makes sense to just do some reasonable default windowing in the main EqualsList function (and then call windowedEquals instead of equals) - if you did that, could you get rid of all your changes to the teststream.go tests?

The advantage of that approach is that users who don't care about explicitly testing windows could just call WindowEquals and wouldn't have to know about this function unless they want a custom windowFn

Reply:
I looked at that a little bit and the issue we hit is that in the PAssert package level we actually don't have any insight into the windowing of the input PCollections. If we could simply pull the windowing information from the primary PCollection out, we'd be able to abstract away all of the extra calls here. I suppose making global windowing the default would work and we just change the underlying equals() function to be what windowedEquals() is now, that may work. Let me iterate on it and make sure that approach isn't breaking. Either way, the teststream tests needed windowing on the PCollections. 

I'm also alright with making this routing standalone because we will eventually want to have robust, streaming-ready passert functions that will need to be windowing aware.

Reply:
> I suppose making global windowing the default would work and we just change the underlying equals() function to be what windowedEquals() is now, that may work

Yeah, something along those lines was my thought - the idea is to make it easier for most people who don't care about windowing-specific equality (those people should use your new function)

> I'm also alright with making this routing standalone because we will eventually want to have robust, streaming-ready passert functions that will need to be windowing aware.

I agree - this is orthogonal to making this standalone or not (I think it should be either way)

Reply:
Routing everything through windowedEquals was fine, done.

Reply:
Ah, I think this actually doesn't quite do what I was hoping for (though it is an improvement IMO), sorry I didn't read your comment closely enough - what I'm hoping for is something that would allow you to leave your teststream test cases unchanged and have it "just work". That would require windowing both the input pcollection and our generated pcollection into some window (e.g. bigWinFn).

Reply:
Oh I see, you want `Equals()` and `EqualsList()` to actually window into a big interval instead of using the global window. I'm not a fan of introducing windowing when it isn't being used explicitly by the user, personally. It isn't intuitive to be evaluating PCollection equality in that way in a global batch context.

We've also found that the current test set-up actually creates a very brittle set-up with a windowed impulse only firing once and not per-window, so this approach doesn't work correctly. PeriodicImpulse and streaming-specific test utilities are really needed.